official/recommendation/neumf_model.py

```diff
@@ -190,37 +190,83 @@ def mlp_slice_fn(x):
     x = tf.squeeze(x, [1])
     return x[:, mf_dim:]
 
-  # It turns out to be significantly more effecient to store the MF and MLP
-  # embedding portions in the same table, and then slice as needed.
-  embedding_user = tf.keras.layers.Embedding(
-      num_users,
-      mf_dim + model_layers[0] // 2,
-      embeddings_initializer=embedding_initializer,
-      embeddings_regularizer=tf.keras.regularizers.l2(mf_regularization),
-      input_length=1,
-      name="embedding_user")(
-          user_input)
-
-  embedding_item = tf.keras.layers.Embedding(
-      num_items,
-      mf_dim + model_layers[0] // 2,
-      embeddings_initializer=embedding_initializer,
-      embeddings_regularizer=tf.keras.regularizers.l2(mf_regularization),
-      input_length=1,
-      name="embedding_item")(
-          item_input)
-
-  # GMF part
-  mf_user_latent = tf.keras.layers.Lambda(
-      mf_slice_fn, name="embedding_user_mf")(embedding_user)
-  mf_item_latent = tf.keras.layers.Lambda(
-      mf_slice_fn, name="embedding_item_mf")(embedding_item)
-
-  # MLP part
-  mlp_user_latent = tf.keras.layers.Lambda(
-      mlp_slice_fn, name="embedding_user_mlp")(embedding_user)
-  mlp_item_latent = tf.keras.layers.Lambda(
-      mlp_slice_fn, name="embedding_item_mlp")(embedding_item)
+  if params["use_tpu"]:
```

Comment:
let's not add this special handling.

Reply:
I don't think there is way out. The two architectures implement embedding differently. With one code path, one would lose performance either for the GPU or the TPU. As I mentioned in the pull request, without special handling, we lose upto 13% performance for the GPU.